File: stacks/Answers/question2/java/stacks.java
/*

2.Check a given word is palindrome or not
	WIRARIW is palindrome


*/

/*implementing stacks using java */

public class main{

    private static String originalWord = "WIRARIW";
    private static int tempSize;
    private static String newWord = "";

    public static void main(String args[]) throws Exception{
        stacks stack1 = new stacks();//create a new stack

        for(int i=0; i<= originalWord.length()-1; i++){
            stack1.push(originalWord.charAt(i));//add original word to the stack
        }

        /* remember that stack1.size() is not a constant, otherwise it will lead to bugs */
  
        tempSize = stack1.size();//temporary store the size of the stack

        for(int i=0; i<= tempSize-1; i++){
            newWord = newWord + stack1.pop();//reversing the word using stack
        }

        if(originalWord.equals(newWord)){
            System.out.println( originalWord + " is a palindrome");
        }else{
            System.out.println( originalWord + " is not a palindrome");
        }
        
    }
} 


class Node{
    private char value;
    private Node nextNode;

    Node(char value){
        this.value = value;
        this.nextNode = null;
    }

    public char getValue() {
        return value;
    }

    public void setValue(char value) {
        this.value = value;
    }

    public Node getNextNode() {
        return nextNode;
    }

    public void setNextNode(Node nextNode) {
        this.nextNode = nextNode;
    }
}


public class stacks{
    private Node top;
    private static int count;
    stacks(){
        this.top = null;
        this.count = 0;
    }

    //Add an value to the stack(to end of the stack)
    public void push(char value){

        if(this.top == null){
            top = new Node(value);//create the first item of the stack
        }else{
            Node node = new Node(value);//create a new node for the stack(top item on the stack)
            node.setNextNode(top);
            top = node;
        }

        count++;
    }

    //Remove the last element from the stack and return the removed value
    public char pop() throws Exception {
        if(this.top == null){
            throw new Exception("Stack is empty");
        }else{
            char value = top.getValue();
            top = top.getNextNode();
            count--;
            return value;

        }

    }

    //Get the size of the stack
    public int size(){
        return count;
    }

    //Get the last element of the stack(but not remove the value)
    public char peek(){
        return top.getValue();
    }

    //check the stack is empty or not
    public boolean isEmpty(){
        if(this.count == 0)
        {
            return true;
        }else{
            return false;
        }

    }


    

}

File: stack/Answers/question2/python/stacks.py
# implementing stacks using python


class Stacks:
    def __init__(self):
        self.count = 0
        self.storage = {}

    # Add an value to the stack(to end of the stack)
    def push(self, value):
        self.storage[self.count] = value
        self.count = self.count + 1

    # Remove the last element from the stack and return the removed value
    def pop(self):
        # if the stack is empty return None
        if self.count == 0:
            raise ValueError('stack is empty')
        else:
            self.count = self.count - 1  # reduce the size of the stack by 1
            pop_value = self.storage[self.count]  # extract the last value from the stack to a variable
            del self.storage[self.count]  # remove the value from the stack
            return pop_value

    # Get the size of the stack
    def size(self):
        return self.count

    # Get the last element of the stack(but not remove the value)
    def peek(self):
        return self.storage[self.count-1]


    # check the stack is empty or not
    def is_empty(self):
        if self.count == 0:
            return True
        else:
            return False

if __name__ == '__main__':

    stack1 = Stacks(); # create a new stack


    originalWord = "WIRARIW"
    newWord = ""

    for i in originalWord:
        stack1.push(i) # add original word to the stack

    # remember that stack1.size() is not a constant, otherwise it will lead to bugs

    tempSize = stack1.size() # temporary store the size of the stack

    for x in range(tempSize):
        newWord = newWord + stack1.pop() # reversing the word using stack

    if(originalWord == newWord):
        print(originalWord + " is a palindrome")
    else:
        print(originalWord + " is not a palindrome")

